From 81d3e57bf3dc42d6dee780ba55e5c4ea90390596 Mon Sep 17 00:00:00 2001
From: Pramoth Suwanpech <pramoth@geniustree.co.th>
Date: Tue, 27 Oct 2026 23:10:19 +0700
Subject: [PATCH] WIP tdf#168666 placeholder - NOT the real commit message

This message is a placeholder so the work can be stored on GitHub.
Per the LibreOffice AI policy, the author must replace it with a commit
message written by himself (git commit --amend) before pushing to Gerrit.

Change-Id: I17db6a9255ec6f0b1611cf7fe7588b3c856b2330
---
 sw/qa/core/txtnode/data/tdf168666.fodt | 203 +++++++++++++++++++++++++
 sw/qa/core/txtnode/txtnode.cxx         |  30 ++++
 sw/source/core/txtnode/fntcache.cxx    |   7 +-
 3 files changed, 238 insertions(+), 2 deletions(-)
 create mode 100644 sw/qa/core/txtnode/data/tdf168666.fodt

diff --git a/sw/qa/core/txtnode/data/tdf168666.fodt b/sw/qa/core/txtnode/data/tdf168666.fodt
new file mode 100644
index 000000000..f75311ee5
--- /dev/null
+++ b/sw/qa/core/txtnode/data/tdf168666.fodt
@@ -0,0 +1,203 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<office:document xmlns:office="urn:oasis:names:tc:opendocument:xmlns:office:1.0" xmlns:style="urn:oasis:names:tc:opendocument:xmlns:style:1.0" xmlns:text="urn:oasis:names:tc:opendocument:xmlns:text:1.0" xmlns:fo="urn:oasis:names:tc:opendocument:xmlns:xsl-fo-compatible:1.0" xmlns:svg="urn:oasis:names:tc:opendocument:xmlns:svg-compatible:1.0" xmlns:loext="urn:org:documentfoundation:names:experimental:office:xmlns:loext:1.0" office:version="1.4" office:mimetype="application/vnd.oasis.opendocument.text">
+ <office:font-face-decls>
+  <!-- Subset of Sarabun (SIL Open Font License 1.1), renamed to avoid clashing with an installed Sarabun -->
+  <style:font-face style:name="Sarabun tdf168666" svg:font-family="'Sarabun tdf168666'" style:font-pitch="variable">
+   <svg:font-face-src>
+    <svg:font-face-uri loext:font-style="normal" loext:font-weight="normal">
+     <office:binary-data>AAEAAAAOAIAAAwBgR0RFRgCZALoAAB/IAAAALEdQT1Ou7ouEAAAf9AAAA5ZHU1VCDM31iQAAI4wA
+AAHwT1MvMl2bZcQAAAFoAAAAYGNtYXC5eoJFAAACbAAAAOJnYXNwAAAAEAAAH8AAAAAIZ2x5ZnL9
+qHEAAAO8AAAXqGhlYWQOQ1XmAAAA7AAAADZoaGVhBB8BuwAAASQAAAAkaG10eCzC7jgAAAHIAAAA
+pGxvY2GpoaNuAAADUAAAAGptYXhwAD8AgAAAAUgAAAAgbmFtZWNViI0AABtkAAAEPHBvc3T/uAAy
+AAAfoAAAACAAAQAAAAEAAENRU15fDzz1AAcD6AAAAADXf927AAAAANe4NEb9av4QAoYE8QAAAAcA
+AgAAAAAAAAABAAAELP8YAAAC0P1q/7QChgABAAAAAAAAAAAAAAAAAAAAHgABAAAANABLAAQANAAE
+AAAAAAAAAAAAAAAAAAAAAgABAAQCFAGQAAUAAAKKAlgAAABLAooCWAAAAV4AMgEsAAAAAAUAAAAA
+AAAAAQAAAQAAAAAAAAAAAAAAAENESyAAwAAgDkwELP8YAAAFBgI3AAAAAQAAAAAB9AK8AAAAIAAD
+ArwAZADzAAACQAArAlIAEQJbACoBxgAVAiwAHQJeABECUAAqAl8AKgKRABcCfQAXAoUAFwKPABcC
+0AAXAmYAJQI8ADUB5AAiAkQAIwIAABICRAAjAnwAFwJLACsCFgAtAeUAHgE2AE4CQwBOAXr/vwGG
+/78AAP6S/gD/Zf9v/q3+nf7N/gD+zv7k/kv+Iv2+/fX9av31/Wr99f1q/fX9av75/vkAAAACAAAA
+AwAAABQAAwABAAAAFAAEAM4AAAAqACAABAAKACAOAg4EDggOCg4VDhcOGw4fDiMOJQ4nDisOLQ4y
+DjgOQQ5EDkkOTP//AAAAIA4BDgQOBw4KDhQOFw4ZDh8OIQ4lDicOKg4tDjAONA5ADkMORw5M////
+4fIB8gDx/vH98fTx8/Hy8e/x7vHt8ezx6vHpAAAAAPHZ8dgAAPHZAAEAAAAAAAAAAAAAAAAAAAAA
+AAAAAAAAAAAAAAAAAA4AEgAAAAAAFgAAAAAAFwAdABgAKgAsAC4AMAAyACgAHwAiAAAAAAAVABUA
+RACbAOkBGAFiAcgCHwJ7ArsDFQNQA4sDywQhBG0EvAUIBUAFmwX5BkEGoQa8BuQG8AdUB4gHvQfw
+B/wICAgVCE8Ihwi/CPYJKQlhCa4J+wohCkcKcgqdCuALIwtUC4ULrAvUAAAAAgBkAAACWALuAAMA
+BwAAEyERISURIRFkAfT+DAHI/mQC7v0SKQKc/WQAAQArAAAB9gJKAB4AADc0NjcmJz4CMzIWFREj
+ETQmIyIGBxYWFxUGBhUVI00nIihDBD5qRGN4VUxBNVEQJEARJitV/CU8Dw4DPV0zblz+gAF4QUwx
+KgQcEhILNyX/AAIAEf/2AggCSgAyAD4AAAERFAYjIiY1NTQ2NzY2NTQmIyIHNjMyFhUUBiMiJjU0
+NjMyFhUUBgcGBhUVFBYzMjY1EQQmIyIGFRQWMzI2NQIIU05OUhMTExUyKDEcBQsnMzQtMDZbUUld
+ExMRECclJSb+8hkUFBkZFBQZAkD+Q0RJSURvGyogHS8cKCYZAjMqKTY/NUNZRUEiMSIcKhlxISMj
+IQHBkxgYFBQZGRQAAAIAKgAAAhECSgAqADYAAAAWFREjETQmIyIGFRQWFxYXNjYzMhYVFAYjIiYn
+BgYVFSM1NCYnJjU0NjMGBhUUFjMyNjU0JiMBkYBVUkdNWQsCAwQQSzYsNDIoFysLFBxUDAcahnMQ
+FxcTExcXEwJKZ1j+dQGDPEZOQhZBCQwYSlsvKigzFRQfXChojw82HF4rYHHjFxMTGBgTExcAAAIA
+FQAAAXwCSgAQABwAAAAWFREjATcXEQYjIiY1NDYzBhYzMjY1NCYjIgYVAT89VP7tL+MPDiw5PC4z
+GxQVGhoVFBsCSjsv/iABLDP2ARoFOS4sOXobGxQUGxsUAAACAB3/9gHiAkoAJwAzAAAEJicnBiMi
+JjU0NjMyFhcXFhYzMjY1ETQmIyIGByM2NjMyFhURFAYjJjY1NCYjIgYVFBYzASxABxQQEyo3OSkk
+NgUkAhkSFhhIPTxSCFUHf2didkNAoRcXExMYGBMKLip5CDQoKTQrIdIPERcUAUAsNDQsS1pWSf6+
+ODv7GBMTFxcTExgAAgAR//YCMAJjAD4ASgAAABUVFAYjIiY1NTQ2NzY2NTQmIyIHNjMyFhUUBiMi
+JjU0NjMyFhUUBgcGBhUVFBYzMjY1NTQmJzU2NjczBgYHBCYjIgYVFBYzMjY1AhRXUE9XExMTFTIo
+MB0FCyczNC0wNltRSFoSEhAPKycnKhseJykCWAQ0H/7MGRQUGRkUFBkBs03jQ0pJRG8bKiAdLxwo
+JhgBMyopNj81Q1lEQiI0HxwqGXEgJCQg5B8kCi8RQDM1UxEdGBgUFBkZFAAAAgAqAAACBgJKADAA
+PAAAABYVESMRNCYjIgYGFRQWFxYWFRU2NzcmJjU0NjMyFhUUBgcGBgcjNTQnJiY1NDY2MwIWMzI2
+NTQmIyIGFQGKfFVPRDRIJAsLCgoVOBciKTMoJzQvMDI1BVQTDQ46b0stGBMTFxcTExgCSmxe/oAB
+eUNJKD4iFjQsKDIUEx8xFgUwIiYyMygmQjAxSi2MFkwwRRs4XTf+5hgYExMXFxMAAgAqAAACFQJK
+ADIAPgAAABYVESMRNCYnBycGBhUUFhcWFhUVNjY3JiY1NDYzMhYVFAYHBgYHIzU0JyYmNTQ2Nxc3
+AhYzMjY1NCYjIgYVAcdOVRodZWkgHgsLCgsPKy8hKTMoJzQwMjQ3BFQTDQ5QSWFijxgTExcXExMY
+AkBtUP59AX8xPhNLTRVCMRY2KCczFCAVKCkELyImMjMoJT4tL0YsjBZKL0QbUHEPTU3+1xgYExIY
+GBIAAAIAFwAAAkcCSgAeACoAAAAVESMRNCMiBgcDIxEGIyImNTQ2MzIWFREUBzcTNjMEFjMyNjU0
+JiMiBhUCR1UhEBEGuFkPDiw5PC4wPQgcZSFU/nIbFBUaGhUUGwJKe/4xAc4zDw/+HQGDBTkuLDk7
+L/76KD5qARJaehsbFBQbGxQAAwAX//YCTwJKACYAMgA+AAAkFhUUBiMiJjU1BgYHIxEGIyImNTQ2
+MzIWFRE2Nz4CNREzERQGBwAWMzI2NTQmIyIGFQA2NTQmIyIGFRQWMwIrJDwuLTopTBdZDw4sOTwu
+MD0bVDc2FFUUE/5CGxQVGhoVFBsBrBoaFRQaGhSvMB4wOzouBBA0HgGDBTkuLDk7L/50HSgaJicc
+AST+2R8yDgEWGxsUFBsbFP5JGxQUGxsUFBsAAgAX//YCOwJKABsAJwAAAREUBiMiJjURBiMiJjU0
+NjMyFhURFBYzMjY1EQQWMzI2NTQmIyIGFQI7b2Jibw8OLDk8LjA9Qjo6Qv5oGxQVGhoVFBsCQP49
+P0hIPwEGBTkuLDk7L/6mIygoIwG6cBsbFBQbGxQAAAIAF//2AkUDLQAbACcAAAERFAYjIiY1EQYj
+IiY1NDYzMhYVERQWMzI2NREAFjMyNjU0JiMiBhUCRXFlYnQPDiw5PC4wPUc6PUT+XhsUFRoaFRQb
+Ay39UD9ISD8BBgU5Liw5Oy/+piIpKCMCp/6jGxsUFBsbFAACABcAAAKGAy0AHAAoAAABESMDAyMR
+BiMiJjU0NjMyFhURFAc3EzMTFyY1EQAWMzI2NTQmIyIGFQKGXZmZXQ8PLDk8LjA9BRVwRW8VBf4c
+GxQVGhoVFBsDLfzTAdj+KAGDBTkuLDk7L/72GTlWAWL+nlY5GQJX/qMbGxQUGxsUAAMAJf/2AhwC
+SgAjAC8AOwAAAREjJiYnFhUUBiMiJjU0NjMyFzUGIyImNTQ2MzIWFREWFhcRBBYzMjY1NCYjIgYV
+EjY1NCYjIgYVFBYzAhxQK2ouAUAxMT0+MgcEDw4sODsvMDw5bCf+lRoUFRsbFRQaUBwcFhUbGxUC
+QP3AJjsOBAgwPT0vLzsBuAU5Liw5Oy/+3RE2HQHncBsbFBQbGxT+TBsWFRsbFRYbAAACADX/9gHy
+AkoAKgA2AAABERQGIyImNTU0NjcmJjU0NjMyFhUUBiMiJxYWMzMVIyIGFRUUFjMyNjURBgYVFBYz
+MjY1NCYjAfJ1ZWV1IiAmJUU6LjYyKyIUASsfXl8aIkc+Pkf3GRkTFBgYFAJA/kpFT09FQiE1DxFL
+LkFOMiopMBAkL0QkHjwoLS0oAbAlGBQTGRkTFBgAAAIAIv/2Ab4CTgAqADYAABIGBxYWFRUUBiMi
+JjU0NjMyFzU0Jic2NjMyFhcWMzI2NxcGBiMiJicmJiMSNjU0JiMiBhUUFjOkJQJwnD0wLzs4LQ4P
+lH4BTUMZLigrDQwUBEAMLh4OLBcGPRVwGhoVFRoaFQIFHx0ORjThLzs5LC45BXkcMRBYXwoLDRQS
+EiouCgYCD/4oGxQUGxsUFBsAAgAj//YB+gJKACoANgAAABYVESM1NCYjIgYVFTYzMhYVFAYjIiY1
+NTQ2NjMyFhc1NCYjIgYHIzY2MwImIyIGFRQWMzI2NQF+fFRWNTVIDw4sODsvMDw2XTkoThtPQ0BW
+CVIHgmgsGhQVGxsVFBoCSl1P/mLrLTQzLS8FOS4sOTsvjDJIJxocSi42NCxMWf4mGxsUFBsbFAAC
+ABL/9gG2AkoAGQAlAAAAFhURFAYjIiY1NDYzMhc1NCYjIgYHIzY2MxI2NTQmIyIGFRQWMwFIbj0w
+Lzs4LQ4PQDkzRwdVB3ZbcBoaFRUaGhUCSl5Q/sQvOzksLjkF1zU8OC5PXP3jGxQUGxsUFBsAAwAj
+//YCFQJ4ACwAMwA/AAAABxYVESM1NCYjIgYVFTYzMhYVFAYjIiY1NTQ2NzY3JiMiBgcjNjYzMhc2
+NzMHNCcGBxYXBiYjIgYVFBYzMjY1Agk4KVRWNTVIDw4sODsvMDxURFI0JC9AVglSB4JoSjgZBWFv
+Diw1Rim+GhQVGxsVFBoCSD8sP/5i5C00My0oBTkuLDk7L4U+Uww8OQ40LExZGScg1xoWKCUJKuEb
+GxQUGxsUAAADABcAAAJFAkoAKQA1AEEAAAAGBxYVESMRNCMiBwMjEQYjIiY1NDYzMhYVERQGBzc2
+NyYmNTQ2MzIWFQQWMzI2NTQmIyIGFSQGFRQWMzI2NTQmIwJFGBUaVRoVFrRLDw4sOTwuMD0IAYIX
+GQ8QOS0tOv4JGxQVGhoVFBsBfBoaFBUaGhUByS0NHzb+xgE6Lib+vgGDBTkuLDk7L/7xITMG7igM
+DicXLDk5LBUbGxQUGxsULxsUFBsbFBQbAAIAK//2AgECSgAnADMAAAAWFREUBiMiJjU1NDYzMhYV
+FAYjIicVFBYzMjY1ETQmIyIGByM2NjMGBhUUFjMyNjU0JiMBhXxuY2NsPC8vOzgsDRBDOzs+T0E/
+VwlSCIVmYhoaFRQaGhQCSl5O/uZDS0tDoC47OC0tOAQ7JyosKAEPLzg0LExZ9BoUFBoaFBQaAAQA
+LQATAdgCNAAVACEANwBDAAASJiY1NDYzMhYVFAcWMzI2NzMOAiMmNjU0JiMiBhUUFjMSJiY1NDYz
+MhYVFAcWMzY2NzMOAiMmNjU0JiMiBhUUFjO4VzQ9LSw8IAYISE0CVAI4aElCGBgVFBgYFCJXND0t
+LDwgBghJTAJUAjhoSUIYGBUUGBgUAUQhPCctOzstKRkBXFNFbT5bGBQVGBgVFBj+dCE8Jy07Oy0p
+GgEDW1JFbT5aGRQUGRkUFBkAAAEAHgAAAZsCSgAPAAABNCYjIgYVIzQ2MzIWFREjAUY3MDI8U2hZ
+V2VVAaoqMTgwUF1VSf5UAAIATv/2ASQCQAANABkAADYWFRQGIyImNREzETYzFjY1NCYjIgYVFBYz
+7Dg7LzA8VA8OFBoaFRUaGhXCOi0sOTsvAeD+fQWVGxQUGxsUFBv//wBO//YCMQJAACIAGQAAAAMA
+GQENAAAAA/+///YBYgPWAC8AOwBHAAAkFhUUBiMiJjURNDY3NjY1NCYjIgYHMzIWFRQGIyImNTQ2
+NjMyFhUUBgcGBhURNjMCJiMiBhUUFjMyNjUSNjU0JiMiBhUUFjMBKTk7LzA9ERALCjEnFyYNByUx
+NSkpNitQNEdgCwwOEBANtRcSEhcXEhIXyBsbFBUaGhXCOi0sOTsvAjIlNCMXHBAiJQ8NMCUnMjUz
+LUkqSD0WIRkbMST+LQQCdRcXEhIXFxL9CBsUFBsbFBQbAAAC/7//9gFuA98AFQAhAAAkFhUUBiMi
+JjURBycmJzMWFzczETYzFjY1NCYjIgYVFBYzATU5PC4wPW9EBCFEFwV+Tw8OExsbFBUaGhXCOi0s
+OTsvAwaaGFWmZ1qy/O0FlRsUFBsbFBQbAAL+kgKnAEwDlwAWACIAABMOAiMiJiY1NDYzMhYVFAYH
+FjMyNjcEFjMyNjU0JiMiBhVMAUBxSDlXMDgrKzgRDwkRUV8B/sMbFBQaGhQUGwOXSG07Ij0mKjY2
+KxcnDQFnUX4aGhQTGhoTAAAC/gACp/+xA5cAFQAhAAADDgIjIiYmNTQ2MzIWFRQHFjMyNjcEFjMy
+NjU0JiMiBhVPAT5tRDlYMDgrLDcgCBBOWgL+zBsUFBsbFBQbA5dIbTsiPSYqNjYrMBsBZVN+GhoU
+ExoaEwAB/2UCsf+1A4cAAwAAAzMVI5tQUAOH1gAB/28D7/+1BLAAAwAAAzMVI5FGRgSwwQAB/q0C
+sf78A4cAAwAAATMVI/6tT08Dh9YAAAL+nQKxAB8DvQAaACYAABMOAiMjNTY2NwYjIiY1NDYzMhYV
+FAYHNjY3BBYzMjY1NCYjIgYVHwldllwmGTARBQolKjAmJjEfGk9rCP77FRAQFRUQERQDsU10PygK
+JBYBLSQjLTAoHkUWC29LVhUVEBAVFBEAAv7NA+8ALwTiABkAJQAAEw4CIyM1NjcGIyImNTQ2MzIW
+FRQGBzY2NwYWMzI2NTQmIyIGFS8JU4dTKDUbBAgiJi0hIy0cF0pgCPATDg8TEw8OEwTXRmk5JBYo
+ASghICksJBs/FAtkRE4TEw8OExMOAAL+AAKx/3wDvQAZACUAAAMOAiMjNTY2NyMiJjU0NjMyFhUU
+Bgc2NjcGFjMyNjU0JiMiBhWECVqTXCYZLREKJiswJicwHxpNZgj+FRAQFRUQEBUDsU10PygKIxYt
+IyQtMCgeRRYLb0tWFRUQERQVEAAC/s4Csf/9A80AFwAjAAADFAYHBgYHBxYWFRQGIyImNTQ2Njc2
+NjUGNjU0JiMiBhUUFjMDLzQJEQccEhQzKCgyLzE9Jid8FhYQERUVEQPNKisTAwUDCQwlFyYyNCko
+MBMUDBsZ7hcRERYWEREXAAL+5APv//sE8QAUACAAAAMUBgcHFhYVFAYjIiY1NDY2NzY2NQY2NTQm
+IyIGFRQWMwUzNyoQEi4kJS4oNCsoJnATExAPExMPBPEsKRENCiEVIywvJSIrFg0MGRnYExAPExMP
+EBMAAAL+SwKx/3oDzQAXACMAAAMUBgcGBwcWFhUUBiMiJjU0NjY3NzY2JwY2NTQmIyIGFRQWM4Yw
+MxQPGxIUMigoMiY1LBYmJwF7FhYREBYWEAPNKS4RBwQJDCUXJjI0KSQuGA4HDBsZ7hcRERYWEREX
+AAAC/iICrP+bBAUAKQA1AAACFhUUBiMiJicGBgcmJjU0NjMzMjczFAYjIyIGFRQWFzY2NxYXJjU0
+NjMWNjU0JiMiBhUUFjOQKi4pI0ghECkQJChPPmVCA0JEQW0dJggHDyoTHScBKiENExIODhISDgNI
+KCEjKyMgCCkXD0YpNUZgSFMjHA0cChIdBR8aBQofJ2oTDQ4TEw4OEgAAAv2+Aqz/MgQFACkANQAA
+AhYVFAYjIiYnBgYHJiY1NDYzMzI3MxQGIyMiBhUUFhc2NjcWFyY1NDYzFjY1NCYjIgYVFBYz+Ssv
+KSNGIA8oECMpTj1iQwNBREFoHSYIBw8pESEhASsgDhISDg4SEg4DSCghIysjIAgoGA9FKjZFYEhT
+IxwNHAoSHQUhFwUJHydqEg4OExMODRMAAAL99QKW/70DgAAKABUAAAImIyIHNjYzMhYXJiYjIgYH
+NjMyFheMvVk+KwV0Xmh/CmNWODJIDBMpP4AwArMfB1Nie299NiggAhQSAAAC/WoClv8pA4AACgAV
+AAAAJiMiBzY2MzIWFyYmIyIGBzYzMhYX/uC4VzwrBXNbZX0KYlM3MUYMFCg+eTACsx8HU2J7b302
+KCACFBIAAv31Apb/vQO4AA4AGQAAAhcmJiMiBzY2MzIXNTMVJhYXJiYjIgYHNjNHBEm9WT4rBXRe
+Yj9I9IAwF1Y4MkgMEykCuCIdHwdTYjdv7zoUEjY2KCACAAAC/WoClv8pA7gADgAZAAACFyYmIyIH
+NjYzMhc1MxUmFhcmJiMiBgc2M9kCSbhXPCsFc1tgPkjveTAWUzcxRgwUKAKsFh0fB1NiOHD4QxQS
+NjYoIAIAAAP99QKW/+EDswAUACAAKwAAAgYHFhcmJiMiBzY2MzIXNjYzMhYVBjY1NCYjIgYVFBYz
+BhYXJiYjIgYHNjMfKSEgBkm9WT4rBXReQjECLCMkLUEVFRAQFRUQz4AwF1Y4MkgMEykDQywCNkkd
+HwdTYhkiKiwjJRUQEBUVEBAVPBQSNjYoIAIAA/1qApb/TgO1ABQAIAArAAACBgcWFyYmIyIHNjYz
+Mhc2NjMyFhUGNjU0JiMiBhUUFjMGFhcmJiMiBgc2M7IoIR4GSbhXPCsFc1tBLwIrIyUsQBQVEBAU
+ExHJeTAWUzcxRgwUKANCLAM2Rx0fB1NiGSMrLSQlFBERFBQRERQ8FBI2NiggAgAC/fUClv+9A7gA
+EwAeAAACFyYmIyIHNjYzMhc1MxUWFzUzFSYWFyYmIyIGBzYzRwRJvVk+KwV0XhsUPSMcPvSAMBdW
+ODJIDBMpArgiHR8HU2IEPE0QHHnvOhQSNjYoIAIAAAL9agKW/ykDuAATAB4AAAIXJiYjIgc2NjMy
+FzUzFRYXNTMVJhYXJiYjIgYHNjPZAkm4VzwrBXNbFhU9Jho+73kwFlM3MUYMFCgCrBYdHwdTYgM7
+TREcevhDFBI2NiggAgAAAv75/rb/tv/EAA0AGQAABwYjIiY1NDYzMhYVFSMmNjU0JiMiBhUUFjOY
+DBEkLjUnLDVODBQUEBAUFBDcBC4kJC4vI7yYFBAQFBQQEBQAAv75/hD/tv7sAA0AGQAAAwYjIiY1
+NDYzMhYVFSMmNjU0JiMiBhUUFjOaDA8kLjUnLDVQChQUEBAUFBD+TAQuJCQuLyOKZhQQEBQUEBAU
+AAAAAA0AogADAAEECQAAAKYAAAADAAEECQABACIApgADAAEECQACAA4AyAADAAEECQADADAA1gAD
+AAEECQAEACIApgADAAEECQAFAEIBBgADAAEECQAGACIBSAADAAEECQAIACoBagADAAEECQAJAFAB
+lAADAAEECQALADQB5AADAAEECQAMAC4CGAADAAEECQANASACRgADAAEECQAOADQDZgBDAG8AcAB5
+AHIAaQBnAGgAdAAgADIAMAAxADgAIABUAGgAZQAgAFMAYQByAGEAYgB1AG4AIABQAHIAbwBqAGUA
+YwB0ACAAQQB1AHQAaABvAHIAcwAgACgAaAB0AHQAcABzADoALwAvAGcAaQB0AGgAdQBiAC4AYwBv
+AG0ALwBjAGEAZABzAG8AbgBkAGUAbQBhAGsALwBTAGEAcgBhAGIAdQBuACkAUwBhAHIAYQBiAHUA
+bgAgAHQAZABmADEANgA4ADYANgA2AFIAZQBnAHUAbABhAHIAUwBhAHIAYQBiAHUAbgAtAHQAZABm
+ADEANgA4ADYANgA2AC0AcwB1AGIAcwBlAHQAVgBlAHIAcwBpAG8AbgAgADEALgAwADAAMAA7ACAA
+dAB0AGYAYQB1AHQAbwBoAGkAbgB0ACAAKAB2ADEALgA2ACkAUwBhAHIAYQBiAHUAbgAtAHQAZABm
+ADEANgA4ADYANgA2AEMAYQBkAHMAbwBuACAARABlAG0AYQBrACAAQwBvAC4ALABMAHQAZAAuAFMA
+dQBwAHAAYQBrAGkAdAAgAEMAaABhAGwAZQByAG0AbABhAHIAcAAgAHwAIABLAGEAdABhAHQAcgBh
+AGQAIABDAG8ALgAsAEwAdABkAC4AaAB0AHQAcAA6AC8ALwB3AHcAdwAuAGMAYQBkAHMAbwBuAGQA
+ZQBtAGEAawAuAGMAbwBtAGgAdAB0AHAAOgAvAC8AdwB3AHcALgBrAGEAdABhAHQAcgBhAGQALgBj
+AG8AbQBUAGgAaQBzACAARgBvAG4AdAAgAFMAbwBmAHQAdwBhAHIAZQAgAGkAcwAgAGwAaQBjAGUA
+bgBzAGUAZAAgAHUAbgBkAGUAcgAgAHQAaABlACAAUwBJAEwAIABPAHAAZQBuACAARgBvAG4AdAAg
+AEwAaQBjAGUAbgBzAGUALAAgAFYAZQByAHMAaQBvAG4AIAAxAC4AMQAuACAAVABoAGkAcwAgAGwA
+aQBjAGUAbgBzAGUAIABpAHMAIABhAHYAYQBpAGwAYQBiAGwAZQAgAHcAaQB0AGgAIABhACAARgBB
+AFEAIABhAHQAOgAgAGgAdAB0AHAAOgAvAC8AcwBjAHIAaQBwAHQAcwAuAHMAaQBsAC4AbwByAGcA
+LwBPAEYATABoAHQAdABwADoALwAvAHMAYwByAGkAcAB0AHMALgBzAGkAbAAuAG8AcgBnAC8ATwBG
+AEwAAwAAAAAAAP+1ADIAAAAAAAAAAAAAAAAAAAAAAAAAAAABAAH//wAPAAEAAAAMAAAAAAAcAAIA
+AgACABYAAQAdADMAAwACAAIAHQAxAAIAMgAzAAEAAQAAAAoALABIAANERkxUABRsYXRuABR0aGFp
+ABQABAAAAAD//wACAAAAAQACbWFyawAObWttawAUAAAAAQAAAAAAAgABAAIAAwAIAdQCGAAEAAAA
+AQAIAAEADAAWAAIAIAB+AAIAAQAdADMAAAACAAEAAgAWAAAAFwABAmgAAQJuAAECXAABAmIAAQJu
+AAECaAABAmIAAQJuAAECaAABAmIAAQJuAAECaAABAm4AAQJoAAECbgABAmgAAQJuAAECaAABAm4A
+AQJoAAECbgAAAcoAAAHQABUAVgBcAGIAaABuAHQAegCAAIYAjACSAJgAngCkAKoAsAC2ALwAwgDI
+AM4A1ADaAOAA5gEoAOwA8gD4AP4BBAEKASIBEAEWARwBIgEoAS4BNAE6AUAAAQH2AAAAAQHsAkAA
+AQH+AAAAAQIIAkAAAQIRAAAAAQIHAkAAAQF8AAAAAQFyAkAAAQHYAAAAAQHYAkAAAQIKAAAAAQIU
+AkAAAQIGAAAAAQH8AkAAAQIVAAAAAQILAkAAAQJHAAAAAQI9AkAAAQIzAAAAAQIzAkAAAQIxAAAA
+AQI7AkAAAQI7AAAAAQG5AkAAAQKGAAAAAQIcAAAAAQIcAkAAAQHoAAAAAQHyAkAAAQF/AAAAAQGa
+AkAAAQHwAkAAAQGsAAAAAQGsAkAAAQH6AAAAAQH6AkAAAQIyAAAAAQIyAkAAAQH3AAAAAQH3AkAA
+BgEAAAEACAABAAwADAABABQAKgABAAIAMgAzAAIAAAAKAAAAEAAB/7YAAAAB/7b/FQACAAYADAAB
+/7b+tgAB/7b+EAAGAgAAAQAIAAEADAAMAAEAFgCEAAIAAQAdADEAAAAVAAAAYgAAAGgAAABWAAAA
+XAAAAGgAAABiAAAAXAAAAGgAAABiAAAAXAAAAGgAAABiAAAAaAAAAGIAAABoAAAAYgAAAGgAAABi
+AAAAaAAAAGIAAABoAAH/tQJAAAH/tgO4AAH/tgJAAAH/KgJAABUALAAyADgAPgBEAEoAUABWAFwA
+YgBoAG4AdAB6AIAAhgCMAJIAmACeAKQAAf+tA5gAAf8TA5gAAf+wA4cAAf+1BLAAAf78A4cAAf+W
+A70AAf/TBOIAAf8mA60AAf+HA80AAf+fBPEAAf8iA80AAf8XBAUAAf6/BAUAAf+2A4oAAf8qA4oA
+Af+1A7gAAf8jA7gAAf+tA7MAAf8ZA7UAAf97A7gAAf7pA7gAAAABAAAACgA4AFQAA0RGTFQAFGxh
+dG4AFHRoYWkAIAAEAAAAAP//AAEAAAAEAAAAAP//AAIAAAABAAJhYWx0AA5jY21wABYAAAACAAAA
+AQAAAAEAAgAIABIAMgBYATQBTgGEAU4BhAABAAAAAQAIAAEABgABAAEABwAdACgAKgAsAC4AMAAy
+AAMAAAABAAgAAQFYAAMADAASABgAAgAgACEAAgAjACQAAgAmACcABgAAAAYAEgBAAFIAbgCQAMIA
+AwABABIAAQEoAAAAAQAAAAMAAgAEAB8AHwAAACIAIgABACUAJQACACgAMQADAAMAAQBIAAEA1gAA
+AAEAAAAEAAMAAQASAAEA6AAAAAEAAAAFAAEAAwAhACQAJwADAAIAFAAaAAEAzAAAAAEAAAAGAAEA
+AQAyAAEAAgANAA4AAwABABIAAQAiAAAAAQAAAAYAAQAGAB4AKQArAC0ALwAxAAEABgAdACgAKgAs
+AC4AMAADAAEAEgABAHgAAAABAAAABwABAAIAHQAeAAEAAAABAAgAAQAGAAEAAQAEAB8AIgAlADIA
+AQAAAAEACAACABgACQAeACEAJAAnACkAKwAtAC8AMQABAAkAHQAfACIAJQAoACoALAAuADAAAQAA
+AAEACAABAAYAAQABAAMAHwAiACU=</office:binary-data>
+     <svg:font-face-format svg:string="truetype"/>
+    </svg:font-face-uri>
+   </svg:font-face-src>
+  </style:font-face>
+ </office:font-face-decls>
+ <office:styles>
+  <style:default-style style:family="paragraph">
+   <style:text-properties style:font-name="Sarabun tdf168666" fo:font-size="14pt" fo:language="en" fo:country="US" style:font-name-complex="Sarabun tdf168666" style:font-size-complex="14pt" style:language-complex="th" style:country-complex="TH"/>
+  </style:default-style>
+ </office:styles>
+ <office:automatic-styles>
+  <style:style style:name="P1" style:family="paragraph">
+   <style:paragraph-properties fo:text-align="justify" style:justify-single-word="false"/>
+  </style:style>
+  <style:page-layout style:name="pm1">
+   <style:page-layout-properties fo:page-width="21.001cm" fo:page-height="29.7cm" fo:margin-top="2cm" fo:margin-bottom="2cm" fo:margin-left="2cm" fo:margin-right="2cm"/>
+  </style:page-layout>
+ </office:automatic-styles>
+ <office:master-styles>
+  <style:master-page style:name="Standard" style:page-layout-name="pm1"/>
+ </office:master-styles>
+ <office:body>
+  <office:text>
+   <text:p text:style-name="P1">การสร้างชุมชนที่แท้จริงคือการยอมเปิดใจรับความเปราะบางของกันและกัน การยอมรับว่าเราต่างก็มีวันที่อ่อนแอและต้องการการประคับประคอง ความเป็นส่วนหนึ่งไม่ได้เกิดจากการที่เรามีไลฟ์สไตล์เหมือนกัน</text:p>
+  </office:text>
+ </office:body>
+</office:document>
diff --git a/sw/qa/core/txtnode/txtnode.cxx b/sw/qa/core/txtnode/txtnode.cxx
index 515df3651..a211c0d47 100644
--- a/sw/qa/core/txtnode/txtnode.cxx
+++ b/sw/qa/core/txtnode/txtnode.cxx
@@ -747,6 +747,36 @@ CPPUNIT_TEST_FIXTURE(SwCoreTxtnodeTest, testPageCrossrefUpdate)
     CPPUNIT_ASSERT_EQUAL(u"2"_ustr, aExpand);
 }
 
+CPPUNIT_TEST_FIXTURE(SwCoreTxtnodeTest, testTdf168666ThaiJustifySmallSpace)
+{
+    // Given a justified Thai paragraph (with an embedded font) where the first line needs less
+    // than 1 twip of extra space per character cell to reach the right margin:
+    createSwDoc("tdf168666.fodt");
+
+    // When painting it:
+    SwDocShell* pShell = getSwDocShell();
+    std::shared_ptr<GDIMetaFile> xMetaFile = pShell->GetPreviewMetaFile();
+    MetafileXmlDump dumper;
+    xmlDocUniquePtr pMetaXml = dumpAndParse(dumper, *xMetaFile);
+
+    // Then the painted first line must be as wide as the print area:
+    xmlDocUniquePtr pLayoutXml = parseLayoutDump();
+    // The first line is a single text portion, so its text array starts at index 0.
+    assertXPath(pLayoutXml,
+                "/root/page[1]/body/txt[1]/SwParaPortion/SwLineLayout[1]/"
+                "SwLinePortion[@type='PortionType::Text']",
+                1);
+    sal_Int32 nLineWidth
+        = getXPath(pLayoutXml, "/root/page[1]/body/txt[1]/infos/prtBounds", "width").toInt32();
+    sal_Int32 nPaintedWidth = getXPath(pMetaXml, "//textarray[1]/dxarray", "last").toInt32();
+    // Without the accompanying fix in place, this test would have failed with:
+    // - Expected: 9638
+    // - Actual  : 9586
+    // i.e. the extra space was truncated to 0 twips per character cell, so the line was painted
+    // without any justification and ended short of the margin.
+    CPPUNIT_ASSERT_DOUBLES_EQUAL(nLineWidth, nPaintedWidth, 2);
+}
+
 CPPUNIT_PLUGIN_IMPLEMENT();
 
 /* vim:set shiftwidth=4 softtabstop=4 expandtab: */
diff --git a/sw/source/core/txtnode/fntcache.cxx b/sw/source/core/txtnode/fntcache.cxx
index 2e665c90d..243984a3f 100644
--- a/sw/source/core/txtnode/fntcache.cxx
+++ b/sw/source/core/txtnode/fntcache.cxx
@@ -1492,7 +1492,9 @@ void SwFntObj::DrawText( SwDrawTextInfo &rInf )
             }
 
             // Thai Justification
-            if ( SwFontScript::CTL == nActual && nSpaceAdd )
+            // tdf#168666 test the precise value: nSpaceAdd is truncated to whole twips, so a line
+            // that needs less than 1 twip per character cell was not justified at all
+            if ( SwFontScript::CTL == nActual && rInf.GetSpace() )
             {
                 LanguageType aLang = rInf.GetFont()->GetLanguage( SwFontScript::CTL );
 
@@ -1979,7 +1981,8 @@ TextFrameIndex SwFntObj::GetModelPositionForViewPoint(SwDrawTextInfo &rInf)
         }
 
         // Thai Justification
-        if ( SwFontScript::CTL == nActual && nSpaceAdd )
+        // tdf#168666 test the precise value, see DrawText()
+        if ( SwFontScript::CTL == nActual && rInf.GetSpace() )
         {
             LanguageType aLang = rInf.GetFont()->GetLanguage( SwFontScript::CTL );
 
-- 
2.56.0

